checkoff

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python3
+import argparse
+import json
+import os
+import subprocess
+import sys
+
+from ocflib.account.search import user_exists
+
+
+# Format of JSON file should be:
+# {
+#     "roster": [<list of OCF usernames of students>],
+#     "users": {
+#         "username1": [<list of labs checked off>],
+#         ...
+#     }
+# }
+
+DATA_PATH = '/home/d/de/decal/checkoff_data.json'
+LOCK_FILE_PATH = DATA_PATH + '.lock'
+
+
+def checkoff(args):
+    # Can replace this with a spin-loop but figured this is alright for now
+    if not _acquire_lock():
+        print('Someone else is doing a checkoff right now, wait a second and rerun your command.')
```

Comment:
Your locking mechanism is quite nice, but I have a suggestion so that multiple runs won't be necessary :)

I'll add comments above lines that may be unclear. 

```python
import fcntl
import os 

with open(DATA_PATH, "r+") as f:
    # Place exclusive lock on this file.
    # If another program has a lock on the 
    # file this will block until it is released.
    fcntl.flock(f, fcntl.LOCK_EX)
    data = json.load(f)
    # Process / modify data.
    ...
    # Go to beginning of the file 
    f.seek(os.SEEK_SET)
    json.dump(data, f)
    # Make file size the new data length
    f.truncate()
    # Release the lock
    fcntl.flock(f, fcntl.LOCK_UN)

Reply:
I was thinking about using `fcntl`, but it would require me to rewrite e.g. `_write_data` and others, since it's very low level working with file pointers and such. I am lazy, so I think this could just be a future improvement.

One thing to watch out for is there might be some instructor that takes like a super long time to approve a checkoff when running the script, in that case having it block would make it confusing for the 2nd instructor who is trying to run the script (they would see no output, script just hangs). This can be fixed with some alert message or timeout of course.